p2p/base_reactor.go

```diff
@@ -68,18 +86,147 @@ type EnvelopeReceiver interface {
 }
 
-func NewBaseReactor(name string, impl Reactor) *BaseReactor {
-	return &BaseReactor{
+type ReactorOptions func(*BaseReactor)
+
+func NewBaseReactor(name string, impl Reactor, opts ...ReactorOptions) *BaseReactor {
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	base := &BaseReactor{
+		ctx:         ctx,
+		cancel:      cancel,
 		BaseService: *service.NewBaseService(nil, name, impl),
 		Switch:      nil,
+		incoming:    make(chan UnprocessedEnvelope, 100),
+		processor:   DefaultProcessor(impl),
+	}
+	for _, opt := range opts {
+		opt(base)
+	}
+
+	go func() {
+		err := base.processor(ctx, base.incoming)
+		if err != nil {
+			err = base.Stop()
+			if err != nil {
+				panic(err)
+			}
+		}
+	}()
+
+	return base
+}
+
+// WithProcessor sets the processor function for the reactor. The processor
+// function is called with the incoming channel and is responsible for
+// unmarshalling the messages and calling Receive on the reactor.
+func WithProcessor(processor ProcessorFunc) ReactorOptions {
+	return func(br *BaseReactor) {
+		br.processor = processor
+	}
+}
+
+// WithIncomingQueueSize sets the size of the incoming message queue for a
+// reactor.
+func WithIncomingQueueSize(size int) ReactorOptions {
+	return func(br *BaseReactor) {
+		br.incoming = make(chan UnprocessedEnvelope, size)
 	}
 }
 
 func (br *BaseReactor) SetSwitch(sw *Switch) {
 	br.Switch = sw
 }
+
+// QueueUnprocessedEnvelope is called by the switch when an unprocessed
+// envelope is received. Unprocessed envelopes are immediately buffered in a
+// queue to avoid blocking. The size of the queue can be changed by passing
+// options to the base reactor.
+func (br *BaseReactor) QueueUnprocessedEnvelope(e UnprocessedEnvelope) {
+	select {
+	// if the context is done, do nothing.
+	case <-br.ctx.Done():
+	// if not, add the item to the channel.
+	case br.incoming <- e:
+	}
+}
+
+func (br *BaseReactor) OnStop() {
+	br.cancel()
+	close(br.incoming)
+}
+
+// DefaultProcessor unmarshalls the message and calls Receive on the reactor.
+// This preserves the sender's original order for all messages.
+func DefaultProcessor(impl Reactor) func(context.Context, <-chan UnprocessedEnvelope) error {
+	implChannels := impl.GetChannels()
+
+	chIDs := make(map[byte]proto.Message, len(implChannels))
+	for _, chDesc := range implChannels {
+		chIDs[chDesc.ID] = chDesc.MessageType
+	}
+	return func(ctx context.Context, incoming <-chan UnprocessedEnvelope) error {
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case ue, ok := <-incoming:
+				if !ok {
+					// this means the channel was closed.
+					return nil
+				}
+				mt := chIDs[ue.ChannelID]
+
+				if mt == nil {
+					return fmt.Errorf("no message type registered for channel %d", ue.ChannelID)
+				}
+
+				msg := proto.Clone(mt)
+
+				err := proto.Unmarshal(ue.Message, msg)
+				if err != nil {
+					return fmt.Errorf("unmarshaling message: %v into type: %s", err, reflect.TypeOf(mt))
```

Comment:
nit: this error message looks confusing b/c the error is included in the middle of the error. Did it mean to say something like:

```suggestion
					return fmt.Errorf("unmarshaling message: %v into type: %s resulted in error %w", msg, reflect.TypeOf(mt), err)
```